File: src/symeraseme/core/scheduler/systemd.py
"""Systemd (Linux) scheduler configuration generator."""

from __future__ import annotations

from symeraseme.core.scheduler.config import SchedulerConfig, _resolve_bin, _resolve_venv
from symeraseme.core.scheduler.wrapper import _wrapper_script


def _systemd_service(
    description: str,
    wrapper_path: str,
    user_service: bool = True,
) -> str:
    return f"""[Unit]
Description={description}
After=network.target

[Service]
Type=oneshot
ExecStart=/bin/bash {wrapper_path}
Environment=SYMERASEME_HEADLESS=1
StandardOutput=journal
StandardError=journal

[Install]
WantedBy=default.target
"""


def _systemd_timer(
    description: str,
    on_calendar: str,
) -> str:
    return f"""[Unit]
Description={description}

[Timer]
OnCalendar={on_calendar}
Persistent=true

[Install]
WantedBy=timers.target
"""


def generate_systemd(
    config: SchedulerConfig,
    project_dir: str = "",
) -> dict[str, str]:
    """Generate systemd service + timer pairs and wrapper scripts."""
    bin_path = config.symeraseme_bin or _resolve_bin()
    venv = config.venv_activate or _resolve_venv()
    out: dict[str, str] = {}

    tick_cmd = f"{bin_path} tick --output json"
    out["symeraseme-tick.sh"] = _wrapper_script(
        tick_cmd, project_dir=project_dir, venv_activate=venv
    )
    out["symeraseme-tick.service"] = _systemd_service(
        "Symaira EraseMe daily tick",
        "__WRAPPER_DIR__/symeraseme-tick.sh",
    )
    out["symeraseme-tick.timer"] = _systemd_timer(
        "Symaira EraseMe daily tick",
        f"Daily-{config.tick_time.hour:02d}:{config.tick_time.minute:02d}",
    )

    poll_cmd = f"{bin_path} poll-inbox --output json"
    out["symeraseme-poll.sh"] = _wrapper_script(
        poll_cmd, project_dir=project_dir, venv_activate=venv
    )
    out["symeraseme-poll.service"] = _systemd_service(
        "Symaira EraseMe hourly inbox poll",
        "__WRAPPER_DIR__/symeraseme-poll.sh",
    )
    out["symeraseme-poll.timer"] = _systemd_timer(
        "Symaira EraseMe hourly inbox poll",
        "Hourly",
    )

    rescan_cmd = f"{bin_path} tick --output json"
    out["symeraseme-rescan.sh"] = _wrapper_script(
        rescan_cmd, project_dir=project_dir, venv_activate=venv
    )
    out["symeraseme-rescan.service"] = _systemd_service(
        "Symaira EraseMe quarterly re-scan",
        "__WRAPPER_DIR__/symeraseme-rescan.sh",
    )
    out["symeraseme-rescan.timer"] = _systemd_timer(
        "Symaira EraseMe quarterly re-scan",
        "Quarterly",
    )

    out["install.sh"] = _systemd_install_script()
    out["uninstall.sh"] = _systemd_uninstall_script()
    return out


def _systemd_install_script() -> str:
    return """#!/usr/bin/env bash
set -euo pipefail
SCRIPT_DIR="$(cd "$(dirname "$0")" && pwd)"
USER_MODE="${1:---user}"

chmod +x "$SCRIPT_DIR"/*.sh

for unit in "$SCRIPT_DIR"/*.service "$SCRIPT_DIR"/*.timer; do
    [ -f "$unit" ] || continue
    name=$(basename "$unit")
    dest="${XDG_CONFIG_HOME:-$HOME/.config}/systemd/user/$name"
    mkdir -p "$(dirname "$dest")"
    sed "s|__WRAPPER_DIR__|$SCRIPT_DIR|g" "$unit" > "$dest"
    echo "Installed: $name"
done

systemctl $USER_MODE daemon-reload
systemctl $USER_MODE enable --now symeraseme-tick.timer
systemctl $USER_MODE enable --now symeraseme-poll.timer
systemctl $USER_MODE enable --now symeraseme-rescan.timer
echo ""
echo "Systemd timers installed and enabled."
echo "Wrappers in: $SCRIPT_DIR"
echo ""
systemctl $USER_MODE list-timers | grep -i symeraseme || true
"""


def _systemd_uninstall_script() -> str:
    return """#!/usr/bin/env bash
set -euo pipefail
USER_MODE="${1:---user}"

for unit in symeraseme-tick symeraseme-poll symeraseme-rescan; do
    systemctl $USER_MODE disable --now "$unit.timer" 2>/dev/null || true
done

for unit in "$SCRIPT_DIR"/*.service "$SCRIPT_DIR"/*.timer; do
    [ -f "$unit" ] || continue
    name=$(basename "$unit")
    dest="${XDG_CONFIG_HOME:-$HOME/.config}/systemd/user/$name"
    rm -f "$dest"
    echo "Removed: $name"
done

systemctl $USER_MODE daemon-reload
echo "Systemd timers uninstalled."
"""

File: crates/symeraseme-engine/src/scheduler/systemd.rs
use super::{Config, WRAPPER_DIR_PLACEHOLDER, wrapper};
use std::collections::BTreeMap;

pub(super) fn generate(
    cfg: &Config,
    binary_path: &str,
    project_dir: &str,
    _poll_hours: &[i32],
) -> BTreeMap<String, String> {
    // systemd invokes the poll wrapper hourly, matching Python; poll hours
    // do not vary the generated unit files.
    let mut files = BTreeMap::new();
    files.insert(
        "symeraseme-tick.sh".to_string(),
        wrapper(
            &[binary_path, "tick", "--output", "json"],
            project_dir,
            &cfg.venv_activate,
        ),
    );
    files.insert(
        "symeraseme-tick.service".to_string(),
        service(
            "Symaira EraseMe daily tick",
            &format!("{WRAPPER_DIR_PLACEHOLDER}/symeraseme-tick.sh"),
        ),
    );
    files.insert(
        "symeraseme-tick.timer".to_string(),
        timer(
            "Symaira EraseMe daily tick",
            "symeraseme-tick",
            &format!("Daily-{:02}:{:02}", cfg.tick_hour, cfg.tick_minute),
        ),
    );
    files.insert(
        "symeraseme-poll.sh".to_string(),
        wrapper(
            &[binary_path, "poll-inbox", "--output", "json"],
            project_dir,
            &cfg.venv_activate,
        ),
    );
    files.insert(
        "symeraseme-poll.service".to_string(),
        service(
            "Symaira EraseMe hourly inbox poll",
            &format!("{WRAPPER_DIR_PLACEHOLDER}/symeraseme-poll.sh"),
        ),
    );
    files.insert(
        "symeraseme-poll.timer".to_string(),
        timer(
            "Symaira EraseMe hourly inbox poll",
            "symeraseme-poll",
            "Hourly",
        ),
    );
    files.insert(
        "symeraseme-rescan.sh".to_string(),
        wrapper(
            &[binary_path, "tick", "--output", "json"],
            project_dir,
            &cfg.venv_activate,
        ),
    );
    files.insert(
        "symeraseme-rescan.service".to_string(),
        service(
            "Symaira EraseMe quarterly re-scan",
            &format!("{WRAPPER_DIR_PLACEHOLDER}/symeraseme-rescan.sh"),
        ),
    );
    files.insert(
        "symeraseme-rescan.timer".to_string(),
        timer(
            "Symaira EraseMe quarterly re-scan",
            "symeraseme-rescan",
            "Quarterly",
        ),
    );
    files.insert("install.sh".to_string(), install_script().to_string());
    files.insert("uninstall.sh".to_string(), uninstall_script().to_string());
    files
}

fn service(description: &str, wrapper_path: &str) -> String {
    format!(
        "[Unit]\n\
         Description={description}\n\
         After=network-online.target\n\
         \n\
         [Service]\n\
         Type=oneshot\n\
         ExecStart=/bin/bash \"{wrapper_path}\"\n\
         Environment=SYMERASEME_HEADLESS=1\n\
         StandardOutput=journal\n\
         StandardError=journal\n\
         \n\
         [Install]\n\
         WantedBy=default.target\n"
    )
}

fn timer(description: &str, service_name: &str, calendar: &str) -> String {
    format!(
        "[Unit]\n\
         Description={description}\n\
         \n\
         [Timer]\n\
         Unit={service_name}.service\n\
         OnCalendar={calendar}\n\
         Persistent=true\n\
         \n\
         [Install]\n\
         WantedBy=timers.target\n"
    )
}

fn install_script() -> &'static str {
    "#!/usr/bin/env bash\n\
     set -euo pipefail\n\
     SCRIPT_DIR=\"$(cd \"$(dirname \"$0\")\" && pwd)\"\n\
     DEST=\"${XDG_CONFIG_HOME:-$HOME/.config}/systemd/user\"\n\
     mkdir -p \"$DEST\"\n\
     chmod +x \"$SCRIPT_DIR\"/*.sh\n\
     for unit in \"$SCRIPT_DIR\"/*.service \"$SCRIPT_DIR\"/*.timer; do\n  \
     [ -f \"$unit\" ] || continue\n  \
     name=$(basename \"$unit\")\n  \
     sed \"s|__WRAPPER_DIR__|$SCRIPT_DIR|g\" \"$unit\" > \"$DEST/$name\"\n\
     done\n\
     systemctl --user daemon-reload\n\
     for timer in symeraseme-tick symeraseme-poll symeraseme-rescan; do\n  \
     systemctl --user enable --now \"$timer.timer\"\n\
     done\n\
     printf '%s\\n' 'Systemd timers installed and enabled.'\n"
}

fn uninstall_script() -> &'static str {
    "#!/usr/bin/env bash\n\
     set -euo pipefail\n\
     SCRIPT_DIR=\"$(cd \"$(dirname \"$0\")\" && pwd)\"\n\
     DEST=\"${XDG_CONFIG_HOME:-$HOME/.config}/systemd/user\"\n\
     for timer in symeraseme-tick symeraseme-poll symeraseme-rescan; do\n  \
     systemctl --user disable --now \"$timer.timer\" 2>/dev/null || true\n\
     done\n\
     for unit in \"$SCRIPT_DIR\"/*.service \"$SCRIPT_DIR\"/*.timer; do\n  \
     [ -f \"$unit\" ] || continue\n  \
     rm -f \"$DEST/$(basename \"$unit\")\"\n\
     done\n\
     systemctl --user daemon-reload\n\
     printf '%s\\n' 'Systemd timers uninstalled.'\n"
}
